docker/autotune.py

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python3
+
+################################################################################
+# Copyright (c) 2023, National Research Foundation (SARAO)
+#
+# Licensed under the BSD 3-Clause License (the "License"); you may not use
+# this file except in compliance with the License. You may obtain a copy
+# of the License at
+#
+#   https://opensource.org/licenses/BSD-3-Clause
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+"""Update the autotuning database."""
+
+import argparse
+import logging
+import os
+import pathlib
+import tempfile
+
+import katsdpsigproc.accel
+
+from katgpucbf.fgpu.ddc import DDCTemplate
+
+
+def main() -> None:  # noqa: D103
+    logging.basicConfig(level=logging.INFO)
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("database", type=pathlib.Path, help="Database filename")
+    args = parser.parse_args()
+
+    # Do all the autotuning in a temporary database, so that the
+    # original is preserved in the event of a failure. Put it in
+    # the same directory so that it can be moved atomicly.
+    with katsdpsigproc.accel.create_some_context(
+        interactive=False,
+        device_filter=lambda device: device.is_cuda,
+    ) as context, tempfile.NamedTemporaryFile(
+        prefix="tuning",
+        suffix=".db",
+        dir=args.database.parent,
+        delete=False,
+    ) as temp_db:
+        should_delete = True
+        try:
+            os.environ["KATSDPSIGPROC_TUNE_DB"] = temp_db.name
```

Comment:
Does the autotune function know to store stuff here but not read from it?

Reply:
It will read from there too, but that's desirable. We want to be isolated from whatever is in the user's existing tuning DB at e.g. `~/.cache/katsdpsigproc/tuning.db`.

Reply:
Right, but presumably since you've just created the file, it'll be empty so the template won't be able to pull values from there, so it knows it'll need to run the autotune.

I think I follow the logic. Sounds reasonable to me.